### PR TITLE
Rust NetDevice and NetDeviceOperationsVtable struct

### PR DESCRIPTION
**Pull request by kloenk**
add function for net_device functions

Contains #226 and #227

### CODE REVIEW

**Comment by kloenk** [comment]
I'm a bit struggling with the `rtnl_link_ops` struct. For the other `*_ops` table I use a const as the FileOperationsVtable. But rtnl_link_ops has to be writable, even if that is not often (`__read_mostly`). Any ideas on this?

Also I'm unsure on how to populate the kind field of that struct

**Comment by adamrk** [comment]
Maybe I'm misunderstanding your issue, but if you're thinking of doing something similar to the existing `FileOperationsVtable` then I'd think it's ok. The `VTABLE` there needs to be const so that it can be constructed at compile time - when the device is registered, a reference to the `file_operations` is passed to the C side and could be mutated.

For the `kind` field, I'd think you'd want to have the user implement a trait with `const KIND: &str` and convert that to the `kind` entry in `rtnl_link_ops`.

**Comment by kloenk** [comment]
@adamrk I think I understand how the `FileOperationsVtable` works. My problem with `rtnl_link_ops` is, that this has to be writable, as (as far as the folks from the kernel told me) there is a linked list inside which is used internally by the kernel. It's only written very rare, but it's written to (that's why it's in `.data.read_mosty`). My Idea is to try it with a macro which creates a `static mut` variable, which than can be assigned to that link. Did not get around yet to implement that, maybe only after the 4th of may as I have to write my finals :-)

**Comment by kloenk** [comment]
I created a proc macro, which creates a wrapped `rtnl_link_ops` (wrapped as we need Sync). This seems to work so far, I'm still trying to reimplement the [drivers/net/dummy](../tree/rust/drivers/net/dummy.c) driver

**Comment by kloenk** [comment]
Rebased to have rust-analyzer, so now contains #226 and #227.

**Comment by kloenk** [comment]
image

I got my first 10 pings via the rust dummy_rs module implemented here. `ip li add name test type dummy_rs` still creates an oops though

**Comment by kloenk** [comment]
and `ip li add name test type dummy_rs` is working now 🎉🎉

Now I have to document everything. help is appreciated :-)

File: drivers/android/rust_binder.rs [hunk 1]
```diff
@@ -35,7 +35,6 @@ module! {
     author: b"Wedson Almeida Filho",
     description: b"Android Binder",
     license: b"GPL v2",
-    params: {},
 }
```
Comment (ojeda) [inline drivers/android/rust_binder.rs]: This should likely go into a different commit -- or is GitHub confused due to having several commits in the PR? Please rebase to have 1 commit anyway.

File: drivers/net/Kconfig [hunk 2]
```diff
@@ -72,6 +72,22 @@ config DUMMY
 	  To compile this driver as a module, choose M here: the module
 	  will be called dummy.
 
+config DUMMY_RS
+	tristate "Dummy net driver support"
+    depends on HAS_RUST
```
Comment (ojeda) [inline drivers/net/Kconfig]: Spaces here -- should be a tab.

File: drivers/net/dummy_rs.rs [hunk 3]
```diff
@@ -0,0 +1,241 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Rust dummy network driver
+//!
+//! This is a demonstration of what a small driver looks like in Rust, based on drivers/net/dummy.c.
+//! This code is provided as a demonstration only, not as a proposal to mass-rewrite existing drivers in Rust
```
Comment (ojeda) [inline drivers/net/dummy_rs.rs:6]: See my comment on https://github.com/Rust-for-Linux/linux/pull/254#discussion_r628663885.

I guess we can talk about this tomorrow in the call :)

File: drivers/net/dummy_rs.rs [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+//! [to go to slip]
+//! 	ifconfig dummy down
+//! 	dip whatever
+//!
+//! This was written by looking at Donald Becker's skeleton driver
+//! and the loopback driver.  I then threw away anything that didn't
+//! apply!	Thanks to Alan Cox for the key clue on what to do with
+//! misguided packets.
+//!
+//!  Finn Behrens, 30th April 20201
```
Comment (ojeda) [inline drivers/net/dummy_rs.rs]: You are in the future! :P

File: drivers/net/dummy_rs.rs [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+//! [when not running slip]
+//! 	ifconfig dummy slip.addr.ess.here up
+//! [to go to slip]
+//! 	ifconfig dummy down
+//! 	dip whatever
+//!
+//! This was written by looking at Donald Becker's skeleton driver
+//! and the loopback driver.  I then threw away anything that didn't
+//! apply!	Thanks to Alan Cox for the key clue on what to do with
+//! misguided packets.
```
Comment (ojeda) [inline drivers/net/dummy_rs.rs]: Hmm... this is a bit confusing now, it sounds like you are talking here since your name is right after the text (rather than coming from `dummy.c`, even if you reference it later).

File: rust/helpers.c [hunk 6]
```diff
@@ -105,8 +108,36 @@ size_t rust_helper_copy_to_iter(const void *addr, size_t bytes, struct iov_iter
...
+EXPORT_SYMBOL_GPL(rust_helper_net_device_set_new_lstats);
+
+void rust_helper_dev_lstats_add(struct net_device *dev, unsigned int len)
+{
+	dev_lstats_add(dev, len);
+}
+EXPORT_SYMBOL_GPL(rust_helper_dev_lstats_add);
+
 // See https://github.com/rust-lang/rust-bindgen/issues/1671
+#if !defined(CONFIG_ARM)
```
Comment (ojeda) [inline rust/helpers.c]: ? Perhaps GitHub is confused -- please rebase.

File: rust/kernel/error.rs [hunk 7]
```diff
@@ -104,3 +107,24 @@ impl From<AllocError> for Error {
         Error::ENOMEM
     }
 }
+
+/// Used by the rtnl_link_ops macro to interface with C
+pub fn c_from_kernel_result<T>(r: KernelResult<T>) -> T
```
Comment (ojeda) [inline rust/kernel/error.rs]: If it is not intended to be used by users, hide the docs.

File: rust/kernel/net/device.rs [hunk 8]
```diff
@@ -0,0 +1,957 @@
...
+        let err = unsafe { bindings::register_netdev(self.ptr as *mut bindings::net_device) };
+
+        if err != 0 {
+            Err(Error::from_kernel_errno(err))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// register_locked - register a network device if the RtnlLock is already hold
```
Comment (ojeda) [inline rust/kernel/net/device.rs]: I guess this is copied from C docs, but do not put the name of the function etc.

File: rust/kernel/net/device.rs [hunk 9]
```diff
@@ -0,0 +1,957 @@
...
+    /// Callers must hold the rtnl semaphore. You may want
+    /// [`register`] instead of this.
+    ///
+    /// BUGS:
+    /// The locking appears insufficient to guarantee two parallel registers
+    /// will not get the same name.
+    ///
+    /// # Safety
+    ///
+    /// caller must hold the RtnlLock and semaphore
```
Comment (ojeda) [inline rust/kernel/net/device.rs:194]:
```suggestion
    /// Caller must hold the [`RtnlLock`] and semaphore.
```

And similar everywhere else.

File: rust/kernel/net/netlink.rs [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+    ///
+    /// # Safety
+    ///
+    /// The pointer `ptr` must be non-null and valid for the lifetime of the object.
+    pub unsafe fn from_ptr(ptr: *const bindings::nlattr) -> Self {
+        Self(ptr)
+    }
+    /*pub unsafe fn from_ptr(ptr: *const bindings::nlattr) -> &'static mut Self {
+        (ptr as *mut NlAttr).as_mut().unwrap()
+    }*/
```
Comment (ojeda) [inline rust/kernel/net/netlink.rs]: I see a bunch of commented out code etc.

**Comment by ojeda** [comment]
> Now I have to document everything. help is appreciated :-)

From the quick review I did above, I assume you are still working on this (i.e. there is a bunch of commented out code, `SAFETY` comments needed, etc.).

File: drivers/net/dummy_rs.rs (hunk 5 above)
Comment (kloenk) [reply drivers/net/dummy_rs.rs]: Oh, yes. Should reword that

File: rust/helpers.c (hunk 6 above)
Comment (kloenk) [reply rust/helpers.c]: Tried that yesterday, but not even the rust branch builds on my system currently. Something with exports.o and symbols

File: drivers/net/dummy_rs.rs (hunk 3 above)
Comment (kloenk) [reply drivers/net/dummy_rs.rs:6]: I think that was the wording of @joshtriplett, while talking about this at our last call.

File: drivers/net/dummy_rs.rs (hunk 5 above)
Comment (kloenk) [reply drivers/net/dummy_rs.rs]: Reworded it like this:
~~~
//! This was written by looking at the dummy network driver from Nick
//! Holloway, which was written by looking at Donald Becker's skeleton driver
//! and the loopback driver

File: rust/helpers.c (hunk 6 above)
Comment (kloenk) [reply rust/helpers.c]: after a `git clean -xfd` (as make mrproper does not work #255 ) it builds the rebased version.
I pushed the rebased version, but I'm sot sure if/why it still makes this as a change.

File: rust/kernel/net/device.rs [hunk 11]
```diff
@@ -0,0 +1,957 @@
...
+        }
+    }
+
+    /// Return const pointer to the wrapped type.
+    pub unsafe fn get_ptr(&self) -> *const bindings::net_device {
+        self.ptr
+    }
+}
+
+impl<I: NetDeviceAdapter> Deref for NetDevice<I> {
```
Comment (kloenk) [inline rust/kernel/net/device.rs]: Should we implement this `Deref` and `DerefMut` or should we consider this unsafe, as this exposes the direct bindings?

File: rust/helpers.c (hunk 6 above)
Comment (ojeda) [reply rust/helpers.c]: Ah, now it is clear -- somehow you ended up with the comment moved outside of the `#ifdef`.

File: drivers/net/dummy_rs.rs (hunk 3 above)
Comment (joshtriplett) [reply drivers/net/dummy_rs.rs:6]: I'm not fixated on the exact wording, I just want to make sure that people don't see a Rust version of an existing C driver and assume we're trying to rewrite the world in Rust. In this context, we're providing samples of what drivers would look like; we're not writing replacements.

In the future, some driver maintainers may choose to use Rust in existing drivers, but that's up to those driver maintainers, not us.

**Comment by kloenk** [comment]
To summarise what I understood in the meeting. I will remove the line saying it is just a test driver. I finish up the abstraction in the `rust/` folder, than we will submit the rust folder, and if it is merged into upstream linux, `I` will submit the `drivers/net/dummy_rs.rs` file myself.
Or would you recommend another way?

**Comment by ojeda** [comment]
If you prefer, you can put it here if you want, too. Like with Binder, for the moment we have it in this tree to have everything in one place. While they will not be merged with the Rust support, I can submit it as an extra patch (with you as the author etc.) like I did with Binder from Wedson in the RFC.

@wedsonaf @TheSven73 since they have the other two potential drivers.

**Comment by kloenk** [comment]
> If you prefer, you can put it here if you want, too. Like with Binder, for the moment we have it in this tree to have everything in one place. While they will not be merged with the Rust support, I can submit it as an extra patch (with you as the author etc.) like I did with Binder from Wedson in the RFC.

Sounds great. I will create a second commit like Wedson explained, so one with abstractions and one with the driver.
Author is not important for me, if it is easier with something else, that's also fine for me. I just looked at the C dummy driver and what is there as author.

**Comment by ojeda** [comment]
If you prefer, different PRs is also fine (and allows us to merge things bit by bit).

**Comment by kloenk** [comment]
I would prefer 2 commits. One pr is easier I guess, but two is also fine. Whatever you prefer.

**Comment by alex** [comment]
FYI: there's a merge conflict here.

**Comment by ksquirrel** [comment]
Review of `f9734782e655`:
  - ⚠️ This PR adds/deletes more than 500 lines.
  - ✔️ Commit [`f973478`](https://github.com/Rust-for-Linux/linux/commit/f9734782e65552a910edf914d8c7955cd12ae06a): Looks fine!

File: rust/kernel/net/ethtool.rs [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+/// Represents which fields of [`struct ethtool_ops`] should pe populated with pointers for the trait [`EthToolOps`]
+pub struct EthToolToUse {
+    /// Trait defines a `get_drvinfo` function.
+    pub get_drvinfo: bool,
+
+    /// Trait defines a `get_ts_info` function.
+    pub get_ts_info: bool,
+}
+
+/// This trait does not include any functions.
```
Comment (soruh) [inline rust/kernel/net/ethtool.rs:138]: These comments should probably be updated to reflect that this struct describes which methods are implemented by an **implementation** of `EthToolOps` not which ones are defined by the trait itself.

File: rust/kernel/net/ethtool.rs [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+                $($i: true),+ ,
+                ..$crate::net::ethtool::ETH_TOOL_USE_NONE
+            };
+    };
+}
+
+/// Operations table needed for ethtool.
+/// [`Self::TO_USE`] defines which functions are implemented by this type.
+pub trait EthToolOps<T: NetDeviceAdapter>: Send + Sync + Sized {
+    /// Struct [`EthToolToUse`] which signals which function are ipmlemeted by this type.
```
Comment (soruh) [inline rust/kernel/net/ethtool.rs:163]: s/ipmlemeted/implemeted/

File: rust/kernel/net/rtnl.rs [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+        unsafe { bindings::rtnl_unlock() };
+    }
+}
+
+#[doc(hidden)]
+/// Empty [`bindings::rtnl_link_ops`] table.
+pub const RTNL_LINK_OPS_EMPTY: bindings::rtnl_link_ops = bindings::rtnl_link_ops {
+    list: bindings::list_head {
+        next: ptr::null::<bindings::list_head>() as *mut bindings::list_head,
+        prev: ptr::null::<bindings::list_head>() as *mut bindings::list_head,
```
Comment (soruh) [inline rust/kernel/net/rtnl.rs:45]: This could use `ptr::null_mut`

File: rust/kernel/net/rtnl.rs [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+        if ret != 0 {
+            Err(Error::from_kernel_errno(ret))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Get a pointer to this struct.
+    pub unsafe fn get_pointer(&self) -> *const bindings::rtnl_link_ops {
+        &self.0 as *const _ as *const bindings::rtnl_link_ops
```
Comment (soruh) [inline rust/kernel/net/rtnl.rs:97]: the `as *const _` seems redundant

File: rust/kernel/net/rtnl.rs [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+    }
+
+    /// Deregister the op table from the kernel.
+    pub fn unregister(&self) {
+        let ptr = self as *const _ as *mut bindings::rtnl_link_ops;
```
Comment (soruh) [inline rust/kernel/net/rtnl.rs:102]: this could be `self.get_pointer() as *mut bindings::rtnl_link_ops`

File: rust/kernel/net/mod.rs [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+/// Return true if the address is valid.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// Please note: addr must be aligned to u16.
+pub unsafe fn is_valid_ether_addr(addr: *const u8) -> bool {
+    !is_multicast_ether_addr(addr) && !is_zero_ether_addr(addr)
+}
```
Comment (soruh) [inline rust/kernel/net/mod.rs:96]: These address functions could generally be made quite a bit safer by using Rust types to ensure proper size/alignment.
I'll type up a draft of what I have in mind if I find the time tomorrow.

File: rust/kernel/net/rtnl.rs [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+impl RtnlLinkStats64 {
+    /// Read the lstats from the [`NetDevice`] `dev` to [`Self`].
+    pub fn dev_read<T: NetDeviceAdapter>(&mut self, dev: &NetDevice<T>) {
+        let stats = self.get_internal_mut();
+        // SAFETY: call to C function
+        unsafe {
+            bindings::dev_lstats_read(
+                dev.get_pointer() as *mut bindings::net_device,
+                &stats.tx_packets as *const u64 as *mut u64,
+                &stats.tx_bytes as *const u64 as *mut u64,
```
Comment (soruh) [inline rust/kernel/net/rtnl.rs:130]: These could be `&mut stats.tx_bytes as *mut u64` instead to clarify that we are not actually making something mutable that isn't.

File: rust/kernel/net/skbuff.rs [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+
+    /// Driver hook for transmit timestamping.
+    ///
+    /// Ethernet MAC Drivers should call this function in their hard_xmit()
+    /// function immediately before giving the sk_buff to the MAC hardware.
+    ///
+    /// Specifically, one should make absolutely sure that this function is
+    /// called before TX completion of this packet can trigger.  Otherwise
+    /// the packet could potentially already be freed.
+    pub fn tx_timestamp(&mut self) {
```
Comment (soruh) [inline rust/kernel/net/skbuff.rs:46]: > Otherwise the packet could potentially already be freed. 

This sounds like this function should potentially be unsafe.

File: rust/kernel/net/skbuff.rs [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+    fn end_pointer(&self) -> *mut u8 {
+        let sk_reff = self.get_internal();
+        (sk_reff.head as usize + sk_reff.end as usize) as *mut u8
+    }
+
+    // !NET_SKBUFF_DATA_USES_OFFSET
+    #[cfg(not(target_pointer_width = "64"))]
+    fn end_pointer(&self) -> *mut u8 {
+        let sk_reff = self.get_internal();
+        (sk_reff.end) as *mut u8
```
Comment (soruh) [inline rust/kernel/net/skbuff.rs:85]: unnecessary parenthesis

File: drivers/net/dummy_rs.rs [hunk 21]
```diff
@@ -0,0 +1,235 @@
...
+        dev.set_new_pcpu_lstats()?;
+        Ok(())
+    }
+
+    fn uninit(dev: &mut NetDevice<Self>) {
+        unsafe { dev.free_lstats() };
+    }
+
+    fn start_xmit(skb: SkBuff, dev: &mut NetDevice<Self>) -> kernel::net::device::NetdevTX {
+        let mut skb = skb;
```
Comment (soruh) [inline drivers/net/dummy_rs.rs:172]: use `fn start_xmit(mut skb: SkBuff, ...` and remove the `let mut skb = skb;`.

File: drivers/net/dummy_rs.rs [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+
+    fn get_stats64(dev: &NetDevice<Self>, stats: &mut rtnl::RtnlLinkStats64) {
+        stats.dev_read(dev);
+    }
+
+    fn change_carrier(dev: &mut NetDevice<Self>, new_carrier: bool) -> Result {
+        dev.carrier_set(new_carrier);
+
+        Ok(())
+    }
```
Comment (soruh) [inline drivers/net/dummy_rs.rs:190]: Why does this return a `Result` if it can't fail?

File: drivers/net/dummy_rs.rs [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+    kernel::declare_eth_tool_ops!(get_drvinfo, get_ts_info);
+
+    fn get_drvinfo(_dev: &NetDevice<Self>, info: &mut ethtool::EthtoolDrvinfo) {
+        // TODO: how to do this more efficient without unsafe?
+        // FIXME: !!
+        let info: &kernel::bindings::ethtool_drvinfo = info.get_internal();
+        unsafe {
+            kernel::bindings::strlcpy(
+                &(info.driver) as *const _ as *mut i8,
+                b"dummy_rs\0" as *const _ as *mut i8,
```
Comment (soruh) [inline drivers/net/dummy_rs.rs:226]: I think you could use the `cstr!` macro here

**Review by soruh — COMMENTED** [review]
There are quite a few places where I'd like more safety comments and the address functions in `net/mod.rs` could probably be safer (see my comment there).
(I didn't look at the changes to `module.rs`)

File: rust/kernel/net/rtnl.rs (hunk 15 above)
Comment (kloenk) [reply rust/kernel/net/rtnl.rs:97]: Yes, but I think rust wants them, as this is not directly the right type for the view of rust

Comment (soruh) [reply rust/kernel/net/rtnl.rs:97]: I think in this case you don't need it is since you're turning a `&rtnl_link_ops` into a `*const rtnl_link_ops`. I agree that the other ones are necessary.

(I could of course be wrong about that and it's ultimately not that important)

Comment (kloenk) [reply rust/kernel/net/rtnl.rs:97]: Could be that I missed .0 befor. And then you need this. Is it is valid, but rust doesn't know that.

Comment (soruh) [reply rust/kernel/net/rtnl.rs:97]: That was probably the case, you can only do `&T as *const T` but not `&U as *const T`, even if `U` only contains a `T`.

Comment (kloenk) [reply rust/kernel/net/rtnl.rs:97]: Yes, as U has repr Transparent. It would also be valid, but rust isn't able to check that. That's the reason to go over *const _

But yeah, will change it.

File: rust/kernel/net/skbuff.rs (hunk 19 above)
Comment (kloenk) [reply rust/kernel/net/skbuff.rs:46]: I think It couldn't, as then we don't have self anymore? But not completely sure right now

Comment (soruh) [reply rust/kernel/net/skbuff.rs:46]: I don't follow completely, are you saying that this is indeed safe/you're not sure about it's safety or that this could't be an unsafe method?

If we're not certain about about the safety of a method, the default should really be making it unsafe.

File: rust/kernel/net/mod.rs [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+/// Return true if the address is all zeroes.
+///
+/// # Parameters
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// Please note: addr must be aligned to u16.
+#[cfg(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)]
+pub unsafe fn is_zero_ether_addr(addr: *const u8) -> bool {
+    *(addr as *const u32) | (*((addr as usize + 4) as *const u16) as u32) == 0
```
Comment (bjorn3) [inline rust/kernel/net/mod.rs]: I am pretty sure this will need to use `ptr::read_unaligned`. Even if the underlying cpu architecture allows efficient unaligned accesses using normal load instructions, LLVM can optimize based on the assumption that `addr as usize % 4 == 0`. For example it could vectorize in ways that would cause overlap if the address is not aligned to 4 bytes. Or it could optimize away manual alignment checks. By the way you can use `addr.offset(4)` instead of `addr as usize + 4`.

File: rust/kernel/net/mod.rs [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+/// - `addr`: Pointer to a six-byte array containing the Ethernet address
+///
+/// # Safety
+/// Please note: addr must be aligned to u16.
+#[cfg(not(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS))]
+pub unsafe fn is_zero_ether_addr(addr: *const u8) -> bool {
+    *(addr as *const u16)
+        | *((addr as usize + 2) as *const u16)
+        | *((addr as usize + 4) as *const u16)
+        == 0
```
Comment (bjorn3) [inline rust/kernel/net/mod.rs:79]: I was thinking that
```suggestion
    *(addr as *const [u16; 3]) == [0; 3]
```
would be nicer to read. When I tested if LLVM would be able to optimize it I noticed the following: On x86_64 and aarch64 the current version uses 3 loads while the suggested version does 2 loads just like the `CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS` version. On armv7 this results in a branch + load from a constant pool for the suggested version while version uses 3 loads and no constant pool.

Because of this I think this suggestion should be used for `CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS` to avoid UB at no perf cost, but `not(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)` should remain the same.

Comment (soruh) [reply rust/kernel/net/mod.rs:79]: Would it maybe make sense for all of these methods to take a `&[u8; 6]` and be completely safe to call or is there an inherent reason we need to treat these values as `u16`s?

Comment (bjorn3) [reply rust/kernel/net/mod.rs:79]: For `CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS` there would be no change. for `not(CONFIG_HAVE_EFFICIENT_UNALIGNED_ACCESS)` that would require 6 loads (one for each u8) instead of 3 (one for each u16).

File: rust/kernel/net/device.rs [hunk 26]
```diff
@@ -0,0 +1,1037 @@
...
+/// The pointer [`NetDevice::ptr`] is non-null and valid.
+#[repr(transparent)]
+pub struct NetDevice<T: NetDeviceAdapter> {
+    ptr: *const bindings::net_device,
+    priv_data: marker::PhantomData<T>,
+}
+
+unsafe impl<T: NetDeviceAdapter> Sync for NetDevice<T> {}
+
+impl<T: NetDeviceAdapter> NetDevice<T> {
```
Comment (nbdd0121) [inline rust/kernel/net/device.rs:61]: Note that Rust's polymorphize support is not yet mature, so one copy of each function might be monomorphized for each `T`. Not a big concern here though.

File: rust/kernel/net/device.rs [hunk 27]
```diff
@@ -0,0 +1,1037 @@
...
+
+    /// This function is called device changes address list filtering.
+    /// If driver handles unicast address filtering, it should set
+    /// IFF_UNICAST_FLT in its priv_flags.
+    fn set_rx_mode(_dev: &mut NetDevice<T>) {}
+}
+
+/// Iff flags
+#[repr(u32)]
+#[allow(non_camel_case_types)]
```
Comment (nbdd0121) [inline rust/kernel/net/device.rs:746]: Any particular reason to use SCREAMING_SNAKE_CASE here?

File: rust/kernel/net/device.rs [hunk 28]
```diff
@@ -0,0 +1,1037 @@
...
+    /// Rename is allowed while device is up and running.
+    LIVE_RENAME_OK = bindings::netdev_priv_flags_IFF_LIVE_RENAME_OK,
+}
+
+/// Feature flags.
+pub mod feature {
+    use core::convert::{From, Into};
+    use core::ops::{Add, AddAssign, Deref, DerefMut, Sub, SubAssign};
+
+    /// Holds multiple flags to give to an interface via [`NetDevice::set_flag`].
```
Comment (nbdd0121) [inline rust/kernel/net/device.rs:863]: This is likely to be very common in bindings. We may want a mini version of `bitflags!`.

File: rust/kernel/net/ethtool.rs [hunk 29]
```diff
@@ -0,0 +1,231 @@
...
+    /// # Safety
+    ///
+    /// The caller must ensure that the adapter is compatible with the way the device is registered.
+    pub(crate) const unsafe fn build() -> &'static bindings::ethtool_ops {
+        &Self::VTABLE
+    }
+}
+
+/// Represents which fields of [`struct ethtool_ops`] should pe populated with pointers for the trait [`EthToolOps`]
+pub struct EthToolToUse {
```
Comment (nbdd0121) [inline rust/kernel/net/ethtool.rs:130]: This also seems to be pretty common for bindings. Maybe we need to have a attribute macro. I'll think about a sensible design.

File: rust/kernel/net/device.rs [hunk 30]
```diff
@@ -0,0 +1,1037 @@
...
+            ptr::write(dest, priv_data);
+        }
+
+        Ok(Self {
+            ptr,
+            priv_data: marker::PhantomData::<T>,
+        })
+    }
+
+    /// Return a refference to the private data of the [`NetDevice`].
```
Comment (nbdd0121) [inline rust/kernel/net/device.rs:122]: s/refference/reference/g

File: rust/kernel/net/device.rs (hunk 28 above)
Comment (bjorn3) [reply rust/kernel/net/device.rs:863]: Excluding tests and docs the `bitflags` crate is a little less than 600 lines big. `bitflags` is a slow moving crate and doesn't implement a huge set of features. I don't know what the policy on vendoring extern crates is, but I don't think there is much benefit of writing a mini version over just vendoring `bitflags`. At most it would save codegen of a couple of methods, though most methods are marked as `#[inline]` and thus only codegened when used.

File: rust/kernel/net/ethtool.rs [hunk 31]
```diff
@@ -0,0 +1,231 @@
...
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the adapter is compatible with the way the device is registered.
+    pub(crate) const unsafe fn build() -> &'static bindings::ethtool_ops {
+        &Self::VTABLE
+    }
+}
+
+/// Represents which fields of [`struct ethtool_ops`] should pe populated with pointers for the trait [`EthToolOps`]
```
Comment (bjorn3) [inline rust/kernel/net/ethtool.rs:129]:
```suggestion
/// Represents which fields of [`struct ethtool_ops`] should be populated with pointers for the trait [`EthToolOps`]
```
nit

**Comment by ksquirrel** [comment]
Review of `818532604b6d`:
  - ⚠️ This PR adds/deletes more than 500 lines.
  - ❌ Commit [`8185326`](https://github.com/Rust-for-Linux/linux/commit/818532604b6d7b8d61389a9c19fccbb5e19f4f08): Private GitHub email address (`@users.noreply.github.com`) used for commit author.
  - ⚠️ Commit [`8185326`](https://github.com/Rust-for-Linux/linux/commit/818532604b6d7b8d61389a9c19fccbb5e19f4f08): The commit committer does not match the commit author.
  - ❌ Commit [`8185326`](https://github.com/Rust-for-Linux/linux/commit/818532604b6d7b8d61389a9c19fccbb5e19f4f08): The last line of the commit message must be a `Signed-off-by` and must match the submitter of the PR.
  - ❌ Commit [`8185326`](https://github.com/Rust-for-Linux/linux/commit/818532604b6d7b8d61389a9c19fccbb5e19f4f08): Unknown tag. Known ones are: `Acked-by`, `Cc`, `Co-developed-by`, `Fixes`, `Link`, `Reported-by`, `Reviewed-by`, `Signed-off-by`, `Suggested-by`, `Tested-by`.

**Comment by kloenk** [comment]
Closing this, as it is quite outdated, and the commit history Is completely broken.
